Hardware Flexibility Tutorial (#386)
stanford-crfm/levanter: docs/Hardware-Agnostic-Training.md

# Switching Hardware Mid Training Run

> One of the standout features of Levanter is its ability to seamlessly operate on both GPU and TPU platforms. This unique feature allows for greater flexibility and adaptability in managing your training resources and is not a capability commonly supported by most frameworks, especially when it involves changing the number or type of accelerators mid-training while preserving the optimizer state. The following guide will walk you through the steps and best practices for switching hardware configurations mid-training run, ensuring a smooth and efficient training process for your models.

Reviewer: The second sentence is too long. I think break it at the `and`. Maybe link with a `Moreover,` ?

### Starting a Training Run

After getting setup up in your [GPU](Getting-Started-GPU.md) or [TPU](Getting-Started-TPU-VM.md) environment (please see [our tutorials](Installation.md) on how to do so if you haven't already), the following command will kick off a training run.

```bash
python src/levanter/main/train_lm.py \
    --config_path config/gpt2_small.yaml \
    --trainer.checkpointer.base_path output/gpt2_small_webtext \
    --trainer.train_batch_size 256
```
This example command assumes you're starting your training run on 2 GPUs in a [FSDP (Fully Sharded Data Parallel)](https://pytorch.org/tutorials/intermediate/FSDP_tutorial.html) setup.

> We have documentation on [getting started with training](Getting-Started-Training.md) in levanter, the [configuration file](Configuration-Guide.md), and [training on your own dataset](Training-On-Your-Data.md) that you should check out for more details. We also have a notebook tutorial on [how to add FSDP to custom architectures](https://colab.research.google.com/drive/1QX4yH3zRFF3Xiibf1aahETcSQ5nbcUMz) implemented in Levanter with Haliax.

Reviewer: ```suggestion
We have documentation on [getting started with training](Getting-Started-Training.md) in Levanter, the [configuration file](Configuration-Guide.md), and [training on your own dataset](Training-On-Your-Data.md) that you should check out for more details. We also have a notebook tutorial on [how to add FSDP to custom architectures](https://colab.research.google.com/drive/1QX4yH3zRFF3Xiibf1aahETcSQ5nbcUMz) implemented in Levanter with Haliax.
```

### How the Model Checkpoint is Saved
The frequency of checkpoint saves can be adjusted by configuring the `checkpointer.save_interval` parameter in your training command. This parameter accepts any reasonable time phrase (15min, 60s, etc), that specifies the interval between each checkpoint save. By default, checkpoints are saved every 15 minutes.

Whenever a checkpoint is saved, Levanter creates a new directory named after the training run ID within the path specified by `trainer.checkpointer.base_path`. This directory stores each checkpoint file. Every checkpoint file containes:

* the model's parameters
* the current optimizer and trainer states
* metadata on the current training step and the time the training step was saved

```
user@DN0a2335e6 rj96a79n/step-1591 % ls
metadata.json	model		training_state	worker-0.cert
```

### Moving the Checkpoint File to Google Storage

> To switch to a new hardware configuration mid training run, you need to move your checkpoint file and your training data over to the new hardware environment. If you started your training run on a GPU server and want to move it to a TPU VM instance, you can use the following command to upload your checkpoint file to a Google Cloud Storage bucket that your TPU VM can access.

Reviewer: ```suggestion
To switch to a new hardware configuration mid-training run, you need to move your checkpoint file and your training data over to the new hardware environment. If you started your training run on a GPU server and want to move it to a TPU VM instance, you can use the following command to upload your checkpoint file to a Google Cloud Storage bucket that your TPU VM can access.
```

```bash
gsutil -m cp -r <GPU MODEL CHECKPT PATH> gs://<SOMEWHERE>
```
You need to be logged into your GCP account in your current environment for this command to run properly. Information on how to set this up can be found [here](Getting-Started-TPU-VM.md).

> ### Re-Configuring `per_device_parallelism`

Reviewer: i think you should retitle this section

# Switching Hardware Mid Training Run

> One of the standout features of Levanter is its ability to seamlessly operate on both GPU and TPU platforms. This guide will walk you through the steps and best practices for switching hardware configurations mid-training, ensuring a smooth and efficient training process for your models.

Reviewer: maybe add a bit here or elsewhere about how changing the number or kind of accelerators midrun (while preserving optimizer state) isn't something most frameworks support?

### Starting a Training Run

After getting setup up in your [GPU](Getting-Started-GPU.md) or [TPU](Getting-Started-TPU-VM.md) environment (please see [our tutorials](Installation.md) on how to do so if you haven't already), the following command will kick off a training run.

```bash
python src/levanter/main/train_lm.py \
    --config_path config/gpt2_small.yaml \
    --trainer.checkpointer.base_path output/gpt2_small_webtext \
    --trainer.train_batch_size 256 \
    --trainer.per_device_parallelism 128
```

> This example command assumes you're starting your training run on 2 GPUs in a [FSDP (Fully Sharded Data Parallel)](https://pytorch.org/tutorials/intermediate/FSDP_tutorial.html) setup. The `trainer.per_device_parallelism` parameter instructs Levanter to distribute the processing of each training batch across multiple accelerators, in this case, across 2 GPUs. With this setting, each GPU handles a subset of the training batch, processing 128 examples each, and their combined efforts contribute to the effective handling of the entire 256 batch size.

Reviewer: you can just leave out `per_device_parallelism` and it'll infer it to be 128. It'll simplify things

> We have comprehensive documentation on [getting started with training](Getting-Started-Training) in levanter, the [configuration file](Configuration-Guide.md), and [training on your own dataset](Training-On-Your-Data.md) that you should check out for more details. We also have a notebook tutorial on [how to add FSDP to custom architectures](https://colab.research.google.com/drive/1QX4yH3zRFF3Xiibf1aahETcSQ5nbcUMz) implemented in Levanter with Haliax.

Reviewer: ```suggestion
We have comprehensive documentation on [getting started with training](Getting-Started-Training.md) in levanter, the [configuration file](Configuration-Guide.md), and [training on your own dataset](Training-On-Your-Data.md) that you should check out for more details. We also have a notebook tutorial on [how to add FSDP to custom architectures](https://colab.research.google.com/drive/1QX4yH3zRFF3Xiibf1aahETcSQ5nbcUMz) implemented in Levanter with Haliax.
```

Reviewer: i also think you can skip the stuff about FSDP in this section, just to keep it simpler?

### How the Model Checkpoint is Saved

> The frequency of checkpoint saves can be adjusted by configuring the `checkpointer.save_interval` parameter in your training command. This parameter requires an integer value, which specifies the interval (in minutes) between each checkpoint save. By default, checkpoints are saved every 15 minutes.

Reviewer: it can actually be any reasonable time phrase (15min, 60s, etc)

Whenever a checkpoint is saved, Levanter creates a new directory named after the training run ID within the path specified by `trainer.checkpointer.base_path`. This directory stores each checkpoint file. Every checkpoint file containes:

* the model's parameters
* the current optimizer and trainer states

> * meta data on the current training step and the time the training step was saved

Reviewer: ```suggestion
* metadata on the current training step and the time the training step was saved
```

```
user@DN0a2335e6 rj96a79n/step-1591 % ls
metadata.json	model		training_state	worker-0.cert
```

### Moving the Checkpoint File to Google Storage

> To switch to a new hardware configuration mid training run, you need to move your checkpoint file and your training data over to the new hardware environemnt. If you started your training run on a GPU server and want to move it to a TPU VM instances, you can use the following command to upload your checkpoint file to a google storage drive that your TPU VM can access.

Reviewer: ```suggestion
To switch to a new hardware configuration mid training run, you need to move your checkpoint file and your training data over to the new hardware environment. If you started your training run on a GPU server and want to move it to a TPU VM instance, you can use the following command to upload your checkpoint file to a Google Cloud Storage bucket that your TPU VM can access.
```

Author: Whoops, I do not have spell check in my VS code. I should get it.

```bash
gsutil -m cp -r <GPU MODEL CHECKPT PATH> gs://<SOMEWHERE>
```
You need to be logged into your GCP account in your current environment for this command to run properly. Information on how to set this up can be found [here](Getting-Started-TPU-VM.md).

> ### Re-Configuring `per_device_parallelism`

Reviewer: again i think you can just leave it unspecified. you need to specify it if you need to do grad accumulation, but otherwise it should "just work"

> When you switch to a new hardware setup, whether its from GPU to TPU or simply agjusting the number of GPUs, it is import you keep your training batch size (`trainer.train_batch_size`) the same. You should however, adjust your `trainer.per_device_parallelism` size to match the new number of accelorators you will be using. For example, if you are moving to a v3-32 TPU VM instances, this means you are using v3 TPUs with 32 TensorCores (I found [this page](https://cloud.google.com/tpu/docs/system-architecture-tpu-vm) helpful for understanding TPU terminology). Each tensor core can handle a portion of your total batch, so you might set your `trainer.per_device_parallelism` size to be 256 / 32 = 8. If you were to set `trainer.per_device_parallelism` to be some suitable value smaller than 8 (like 1, 2, or 4), then Levanter would perform gradient accumlation until 8 examples have been processed by the model on this accelorator.

Reviewer: ```suggestion
When you switch to a new hardware setup, whether its from GPU to TPU or simply adjusting the number of GPUs, it is important you keep your training batch size (`trainer.train_batch_size`) the same. You should however, adjust your `trainer.per_device_parallelism` size to match the new number of accelerators you will be using. For example, if you are moving to a v3-32 TPU VM instances, this means you are using v3 TPUs with 32 TensorCores (I found [this page](https://cloud.google.com/tpu/docs/system-architecture-tpu-vm) helpful for understanding TPU terminology). Each tensor core can handle a portion of your total batch, so you might set your `trainer.per_device_parallelism` size to be 256 / 32 = 8. If you were to set `trainer.per_device_parallelism` to be some suitable value smaller than 8 (like 1, 2, or 4), then Levanter would perform gradient accumulation until 8 examples have been processed by the model on this accelerator.
```

Reviewer: I'm not totally sure it is important to keep the train batch size the same, but it's probably a good idea. changing the number of train steps is probably asking for trouble b/c it messes with the LR schedule

## Resuming A Training Run
Once your new hardware environment is setup and you've moved your training checkpoint over, you can resume your training run with this command:

```bash
python levanter/src/levanter/main/train_lm.py \
	--config_path levanter/config/gpt2_small.yaml \
	--trainer.load_checkpoint_path gs://<SOMEWHERE> \
	--trainer.wandb.resume true \
	--trainer.id rj96a79n
```
Use `trainer.load_checkpoint_path` to specify where the model checkpoint you moved is located. `trainer.wandb.resume true` will resume all your learning curves where the stopped training run left off, and `trainer.id` is the trainer ID used by WandB to identify your stopped training run.

## Picking Up Where You Left Off

> When your training run resumes, the traininer will first fast-forward through your training data to continue from the exact point in the epoch where your checkpoint was saved. You should see a progress bar tracking the fast-forward through your data.

Reviewer: ```suggestion
When your training run resumes, the trainer will first fast-forward through your training data to continue from the exact point in the epoch where your checkpoint was saved. You should see a progress bar tracking the fast-forward through your data.
```

> Resuming your training run restores your checkpoint's optimizers states, so your WandB learning curves should resume smoothly from where they left off.

Reviewer: ```suggestion
Resuming your training run restores your checkpoint's optimizer states, so your WandB learning curves should resume smoothly from where they left off.
```
